File: python/l1_pager/schema.py
"""
L1-Pager: Core schema for the Context Garbage Collector.

Design invariant: every type here must be JSON-serializable so the
Python runtime and the TypeScript runtime can share the same Redis heap
without a translation layer.
"""
from __future__ import annotations

import hashlib
import json
import re
import time
from dataclasses import dataclass, field, asdict
from enum import Enum
from typing import Annotated, Any, Literal

from langchain_core.messages import AIMessage, BaseMessage, ToolMessage
from langgraph.graph.message import add_messages
from typing_extensions import TypedDict

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

PAGE_POINTER_PREFIX = "<PAGE_FAULT_ID:"
_POINTER_RE = re.compile(r"<PAGE_FAULT_ID:\s*([a-f0-9]+)\s*\|\s*SUMMARY:\s*(.+?)>$")

# ---------------------------------------------------------------------------
# Eviction policy
# ---------------------------------------------------------------------------


class EvictionPolicy(str, Enum):
    LRU = "lru"
    IMPORTANCE = "importance"
    # Weighted blend: importance_score - (turn_age * decay_weight)
    HYBRID = "hybrid"


@dataclass(frozen=True)
class EvictionConfig:
    """All knobs for the eviction engine. Immutable after construction."""

    # Mark criteria
    min_tokens: int = 500
    min_turns_old: int = 3

    # Sweep behaviour
    policy: EvictionPolicy = EvictionPolicy.LRU
    hybrid_decay_weight: float = 0.01

    # Heap limits
    max_heap_entries: int = 2_000
    entry_ttl_seconds: float = 3_600.0

    # Demand-paging guard
    max_page_fault_depth: int = 3

    def to_dict(self) -> dict:
        return {
            "min_tokens": self.min_tokens,
            "min_turns_old": self.min_turns_old,
            "policy": self.policy.value,
            "hybrid_decay_weight": self.hybrid_decay_weight,
            "max_heap_entries": self.max_heap_entries,
            "entry_ttl_seconds": self.entry_ttl_seconds,
            "max_page_fault_depth": self.max_page_fault_depth,
        }

    @classmethod
    def from_dict(cls, d: dict) -> EvictionConfig:
        d = dict(d)
        d["policy"] = EvictionPolicy(d["policy"])
        return cls(**d)


# ---------------------------------------------------------------------------
# Heap entry — the raw content stored after eviction
# ---------------------------------------------------------------------------


@dataclass
class PageHeapEntry:
    """
    Stored in Redis/InMemory after a message is swept.

    `raw_content` is the original message.content value. It must be
    JSON-serializable; complex content blocks (list[dict]) satisfy this
    by construction in LangChain.
    """

    page_id: str                    # 16-char hex prefix of SHA-256
    original_message_id: str        # LangGraph message id (for re-injection)
    original_type: str              # "ToolMessage" | "AIMessage" | …
    raw_content: Any                # str or list[dict] — the actual payload
    tool_name: str | None           # Non-None for ToolMessage
    tool_call_id: str | None        # Non-None for ToolMessage
    token_count: int
    evicted_at_turn: int
    evicted_at_ts: float = field(default_factory=time.time)
    last_accessed_ts: float | None = None
    access_count: int = 0

    # ---- serialisation ----

    def to_dict(self) -> dict:
        return {
            "page_id": self.page_id,
            "original_message_id": self.original_message_id,
            "original_type": self.original_type,
            "raw_content": self.raw_content,
            "tool_name": self.tool_name,
            "tool_call_id": self.tool_call_id,
            "token_count": self.token_count,
            "evicted_at_turn": self.evicted_at_turn,
            "evicted_at_ts": self.evicted_at_ts,
            "last_accessed_ts": self.last_accessed_ts,
            "access_count": self.access_count,
        }

    @classmethod
    def from_dict(cls, d: dict) -> PageHeapEntry:
        return cls(**d)

    def touch(self) -> None:
        """Update LRU metadata on access."""
        self.last_accessed_ts = time.time()
        self.access_count += 1

    @property
    def lru_ts(self) -> float:
        return self.last_accessed_ts or self.evicted_at_ts


# ---------------------------------------------------------------------------
# Mark candidate — output of the Mark phase
# ---------------------------------------------------------------------------


@dataclass
class MarkCandidate:
    """A message flagged by the Mark phase as eligible for eviction."""

    message: BaseMessage
    message_index: int       # Position in the current messages list
    token_count: int
    turn_age: int            # Number of full turns since this message
    importance_score: float  # [0.0, 1.0] — lower = evict sooner
    eviction_reason: str


# ---------------------------------------------------------------------------
# Eviction record — audit log entry
# ---------------------------------------------------------------------------


@dataclass
class EvictionRecord:
    page_id: str
    original_message_id: str
    original_type: str
    token_count: int
    evicted_at_turn: int
    evicted_at_ts: float = field(default_factory=time.time)
    policy_used: str = "lru"

    def to_dict(self) -> dict:
        return asdict(self)


# ---------------------------------------------------------------------------
# Pointer helpers
# ---------------------------------------------------------------------------


def compute_page_id(message: BaseMessage) -> str:
    """
    Deterministic, stable 16-char hex page ID.

    Keyed on (message.id, type, content) so the same logical message
    always maps to the same heap slot — safe to call multiple times.
    """
    payload = json.dumps(
        {
            "id": message.id,
            "type": type(message).__name__,
            "content": message.content,
        },
        sort_keys=True,
        default=str,
    )
    return hashlib.sha256(payload.encode()).hexdigest()[:16]


def render_pointer(page_id: str, summary: str) -> str:
    return f"<PAGE_FAULT_ID: {page_id} | SUMMARY: {summary}>"


def parse_pointer(content: Any) -> tuple[str, str] | None:
    """Return (page_id, summary) if `content` is a pointer string, else None."""
    if not isinstance(content, str):
        return None
    m = _POINTER_RE.match(content.strip())
    return (m.group(1), m.group(2)) if m else None


def is_pointer_content(content: Any) -> bool:
    return parse_pointer(content) is not None


def make_page_pointer_message(
    original: BaseMessage,
    page_id: str,
    summary: str,
    evicted_at_turn: int,
) -> BaseMessage:
    """
    Build a pointer message that replaces `original` in the message array.

    Critically, the returned message carries the *same id* as `original`
    so LangGraph's add_messages reducer performs an in-place update rather
    than appending a duplicate.
    """
    pointer_content = render_pointer(page_id, summary)
    extra_kwargs: dict = {
        "l1_page_id": page_id,
        "l1_original_type": type(original).__name__,
        "l1_evicted_at_turn": evicted_at_turn,
    }
    original_extra: dict = getattr(original, "additional_kwargs", {}) or {}

    if isinstance(original, ToolMessage):
        return ToolMessage(
            id=original.id,
            content=pointer_content,
            tool_call_id=original.tool_call_id,
            name=original.name,
            additional_kwargs={**original_extra, **extra_kwargs},
        )

    if isinstance(original, AIMessage):
        return AIMessage(
            id=original.id,
            content=pointer_content,
            additional_kwargs={**original_extra, **extra_kwargs},
        )

    # Fallback: reconstruct same type with updated content
    try:
        return type(original)(
            id=original.id,
            content=pointer_content,
            additional_kwargs={**original_extra, **extra_kwargs},
        )
    except Exception:
        return AIMessage(
            id=original.id,
            content=pointer_content,
            additional_kwargs={**original_extra, **extra_kwargs},
        )


# ---------------------------------------------------------------------------
# L1PagerState — optional TypedDict mixin for full observability
# ---------------------------------------------------------------------------


class L1PagerState(TypedDict, total=False):
    """
    State mixin for graphs that want L1Pager to be first-class in state.

    Usage::

        class MyState(L1PagerState):
            messages: Annotated[list[BaseMessage], add_messages]
            my_field: str

    If you do NOT include this mixin, L1Pager still works — it stores
    metadata in the external heap and a thread-local cache.
    """

    messages: Annotated[list[BaseMessage], add_messages]
    l1_turn_count: int
    # page_id -> PageHeapEntry.to_dict() — survives graph checkpointing
    l1_heap_index: dict[str, dict]
    l1_eviction_log: list[dict]
    # Guards recursive demand-paging from infinite loops
    l1_page_fault_depth: int

File: typescript/src/schema.ts
/**
 * L1-Pager: Core schema for the Context Garbage Collector.
 *
 * All types here are JSON-serializable so the TS runtime and the Python
 * runtime can share the same Redis heap key-space without translation.
 */

import {
  AIMessage,
  BaseMessage,
  MessageContent,
  ToolMessage,
} from "@langchain/core/messages";
import { Annotation, messagesStateReducer } from "@langchain/langgraph";
import { createHash } from "crypto";

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

export const PAGE_POINTER_PREFIX = "<PAGE_FAULT_ID:" as const;
const POINTER_RE = /^<PAGE_FAULT_ID:\s*([a-f0-9]+)\s*\|\s*SUMMARY:\s*(.+?)>$/;

// ---------------------------------------------------------------------------
// Eviction policy
// ---------------------------------------------------------------------------

export type EvictionPolicy = "lru" | "importance" | "hybrid";

export interface EvictionConfig {
  /** Minimum estimated tokens for a message to be eviction-eligible. */
  readonly minTokens: number;
  /** Minimum turns-old a message must be to be eviction-eligible. */
  readonly minTurnsOld: number;
  readonly policy: EvictionPolicy;
  /** Weight applied to turn age in the hybrid policy. */
  readonly hybridDecayWeight: number;
  readonly maxHeapEntries: number;
  readonly entryTtlSeconds: number;
  /** Guard: maximum recursive depth for demand-paging page faults. */
  readonly maxPageFaultDepth: number;
}

export const DEFAULT_EVICTION_CONFIG: EvictionConfig = {
  minTokens: 500,
  minTurnsOld: 3,
  policy: "lru",
  hybridDecayWeight: 0.01,
  maxHeapEntries: 2_000,
  entryTtlSeconds: 3_600,
  maxPageFaultDepth: 3,
};

// Wire Python <-> TS: field names use snake_case in the wire format
export function evictionConfigToWire(cfg: EvictionConfig): Record<string, unknown> {
  return {
    min_tokens: cfg.minTokens,
    min_turns_old: cfg.minTurnsOld,
    policy: cfg.policy,
    hybrid_decay_weight: cfg.hybridDecayWeight,
    max_heap_entries: cfg.maxHeapEntries,
    entry_ttl_seconds: cfg.entryTtlSeconds,
    max_page_fault_depth: cfg.maxPageFaultDepth,
  };
}

export function evictionConfigFromWire(raw: Record<string, unknown>): EvictionConfig {
  return {
    minTokens: raw["min_tokens"] as number,
    minTurnsOld: raw["min_turns_old"] as number,
    policy: raw["policy"] as EvictionPolicy,
    hybridDecayWeight: raw["hybrid_decay_weight"] as number,
    maxHeapEntries: raw["max_heap_entries"] as number,
    entryTtlSeconds: raw["entry_ttl_seconds"] as number,
    maxPageFaultDepth: raw["max_page_fault_depth"] as number,
  };
}

// ---------------------------------------------------------------------------
// PageHeapEntry — raw content stored in the heap after eviction
// ---------------------------------------------------------------------------

/**
 * Wire format (snake_case) so both Python and TS agents read/write
 * the same Redis hash without a translation step.
 */
export interface PageHeapEntryWire {
  page_id: string;
  original_message_id: string;
  original_type: string;
  raw_content: MessageContent;
  tool_name: string | null;
  tool_call_id: string | null;
  token_count: number;
  evicted_at_turn: number;
  evicted_at_ts: number;
  last_accessed_ts: number | null;
  access_count: number;
}

export class PageHeapEntry {
  pageId: string;
  originalMessageId: string;
  originalType: string;
  rawContent: MessageContent;
  toolName: string | null;
  toolCallId: string | null;
  tokenCount: number;
  evictedAtTurn: number;
  evictedAtTs: number;
  lastAccessedTs: number | null;
  accessCount: number;

  constructor(fields: {
    pageId: string;
    originalMessageId: string;
    originalType: string;
    rawContent: MessageContent;
    toolName?: string | null;
    toolCallId?: string | null;
    tokenCount: number;
    evictedAtTurn: number;
    evictedAtTs?: number;
    lastAccessedTs?: number | null;
    accessCount?: number;
  }) {
    this.pageId = fields.pageId;
    this.originalMessageId = fields.originalMessageId;
    this.originalType = fields.originalType;
    this.rawContent = fields.rawContent;
    this.toolName = fields.toolName ?? null;
    this.toolCallId = fields.toolCallId ?? null;
    this.tokenCount = fields.tokenCount;
    this.evictedAtTurn = fields.evictedAtTurn;
    this.evictedAtTs = fields.evictedAtTs ?? Date.now() / 1_000;
    this.lastAccessedTs = fields.lastAccessedTs ?? null;
    this.accessCount = fields.accessCount ?? 0;
  }

  get lruTs(): number {
    return this.lastAccessedTs ?? this.evictedAtTs;
  }

  touch(): void {
    this.lastAccessedTs = Date.now() / 1_000;
    this.accessCount += 1;
  }

  toWire(): PageHeapEntryWire {
    return {
      page_id: this.pageId,
      original_message_id: this.originalMessageId,
      original_type: this.originalType,
      raw_content: this.rawContent,
      tool_name: this.toolName,
      tool_call_id: this.toolCallId,
      token_count: this.tokenCount,
      evicted_at_turn: this.evictedAtTurn,
      evicted_at_ts: this.evictedAtTs,
      last_accessed_ts: this.lastAccessedTs,
      access_count: this.accessCount,
    };
  }

  static fromWire(w: PageHeapEntryWire): PageHeapEntry {
    return new PageHeapEntry({
      pageId: w.page_id,
      originalMessageId: w.original_message_id,
      originalType: w.original_type,
      rawContent: w.raw_content,
      toolName: w.tool_name,
      toolCallId: w.tool_call_id,
      tokenCount: w.token_count,
      evictedAtTurn: w.evicted_at_turn,
      evictedAtTs: w.evicted_at_ts,
      lastAccessedTs: w.last_accessed_ts,
      accessCount: w.access_count,
    });
  }
}

// ---------------------------------------------------------------------------
// MarkCandidate — output of the Mark phase
// ---------------------------------------------------------------------------

export interface MarkCandidate {
  message: BaseMessage;
  messageIndex: number;
  tokenCount: number;
  turnAge: number;
  importanceScore: number;
  evictionReason: string;
}

// ---------------------------------------------------------------------------
// EvictionRecord — audit log entry
// ---------------------------------------------------------------------------

export interface EvictionRecord {
  pageId: string;
  originalMessageId: string;
  originalType: string;
  tokenCount: number;
  evictedAtTurn: number;
  evictedAtTs: number;
  policyUsed: EvictionPolicy;
}

// ---------------------------------------------------------------------------
// Pointer helpers
// ---------------------------------------------------------------------------

export function computePageId(message: BaseMessage): string {
  const payload = JSON.stringify(
    {
      id: message.id,
      type: message._getType(),
      content: message.content,
    },
    (_, v) => (v === undefined ? null : v)
  );
  return createHash("sha256").update(payload).digest("hex").slice(0, 16);
}

export function renderPointer(pageId: string, summary: string): string {
  return `<PAGE_FAULT_ID: ${pageId} | SUMMARY: ${summary}>`;
}

export function parsePointer(content: unknown): { pageId: string; summary: string } | null {
  if (typeof content !== "string") return null;
  const m = POINTER_RE.exec(content.trim());
  return m ? { pageId: m[1], summary: m[2] } : null;
}

export function isPointerContent(content: unknown): boolean {
  return parsePointer(content) !== null;
}

/**
 * Build a replacement message carrying the same `id` as `original`.
 *
 * LangGraph's messagesStateReducer deduplicates by id, so returning this
 * message from a node produces an in-place update — not an append.
 */
export function makePagePointerMessage(
  original: BaseMessage,
  pageId: string,
  summary: string,
  evictedAtTurn: number
): BaseMessage {
  const pointerContent = renderPointer(pageId, summary);
  const extraKwargs = {
    l1_page_id: pageId,
    l1_original_type: original._getType(),
    l1_evicted_at_turn: evictedAtTurn,
  };
  const originalKwargs = (original.additional_kwargs as Record<string, unknown>) ?? {};

  if (original instanceof ToolMessage) {
    return new ToolMessage({
      id: original.id,
      content: pointerContent,
      tool_call_id: original.tool_call_id,
      name: original.name,
      additional_kwargs: { ...originalKwargs, ...extraKwargs },
    });
  }

  // AIMessage and everything else
  return new AIMessage({
    id: original.id,
    content: pointerContent,
    additional_kwargs: { ...originalKwargs, ...extraKwargs },
  });
}

// ---------------------------------------------------------------------------
// L1PagerState — LangGraph Annotation (opt-in mixin)
// ---------------------------------------------------------------------------

/**
 * Optional state annotation for graphs that want full L1Pager observability.
 *
 * Usage:
 *   const MyState = Annotation.Root({
 *     ...L1PagerAnnotation.spec,
 *     myField: Annotation<string>({ default: () => "" }),
 *   });
 */
export const L1PagerAnnotation = Annotation.Root({
  messages: Annotation<BaseMessage[]>({
    reducer: messagesStateReducer,
    default: () => [],
  }),
  l1TurnCount: Annotation<number>({
    reducer: (prev, next) => next,
    default: () => 0,
  }),
  /** page_id -> PageHeapEntryWire — survives graph checkpointing */
  l1HeapIndex: Annotation<Record<string, PageHeapEntryWire>>({
    reducer: (prev, next) => ({ ...prev, ...next }),
    default: () => ({}),
  }),
  l1EvictionLog: Annotation<EvictionRecord[]>({
    reducer: (prev, next) => [...prev, ...next],
    default: () => [],
  }),
  /** Guards recursive demand-paging from infinite loops. */
  l1PageFaultDepth: Annotation<number>({
    reducer: (prev, next) => next,
    default: () => 0,
  }),
});

export type L1PagerState = typeof L1PagerAnnotation.State;
